### ISSUE TITLE
accepts_nested_attributes_for breaks the uniqueness with scope validation

### ISSUE THREAD

**Issue by spurnaye** [rails/rails#20676]
class Country < ActiveRecord::Base
  has_many :cities, dependent: :destroy
  accepts_nested_attributes_for :cities
end

class Translation < ActiveRecord::Base
  belongs_to :source
  validates :text, presence: true
  validates :language, presence: true, locale: true
  validates :language , uniqueness: { scope: :source_id }
#  validates :source_id
end

**Comment by meinac** [comment]
@spurnaye As far as I can see from your script you don't need to use `accepts_nested_attributes_for` for saving cities with country like at the end of your script. 
I've tested this with `master`, `4.2.2` and `4.1.0` the behaviour is same in all versions, your test case fails. I think this is a bug about autosaving associations not about `accepts_nested_attributes_for`. I'll dig into this.

**Comment by al2o3cr** [comment]
@meinac You're correct that `accepts_nested_attributes_for` is not needed for this use case, but it's relevant because it changes the behavior of autosaving a collection slightly.

Without it, records in the collection are validated when being saved with a new parent record. *With* it, those validations are explicitly skipped:

https://github.com/rails/rails/blob/master/activerecord/lib/active_record/autosave_association.rb#L377

The logic is a little strange to read because there's a difference in behavior between "not specifying `autosave` in the association at all` vs. "explicitly setting it to `false`".

I'd also point out that if uniqueness is an important requirement for you that the first example shouldn't "work" even when AR behaves unexpectedly - you should have a DB constraint to enforce that in a race-condition-free way.

**Comment by meinac** [comment]
@al2o3cr You are right `accepts_nested_attributes` sets autosave to true therefore rails does not validates collections with it. 
>The logic is a little strange to read because there's a difference in behavior between "not specifying autosave in the association at allvs. "explicitly setting it tofalse`".

I agree with you. It looks really strange, as far as I can understand from the code, in case of explicitly set as `true` `autosave` case it rollbacks the transaction if can not save the record. 

I had worked on a bug of autosave and I've created a pull request as a POC in #19724. I believe we should refactor the logic of autosave.

**Comment by spurnaye** [comment]
@meinac and @al2o3cr I tested with 3-2-stable branch without the accepts_nested_attributes_for and see that the issue persists with validation. Testing further with latest versions

**Comment by meinac** [comment]
I've found the problem. As I said before the problem is in autosave associations. `UniquenessValidator` works as expected it adds error to the second city record but in autosave associations active record ignores this and discards the record. At the end of the transaction it inserts country record and only one city record into the database. 
The problem is at `association.insert_record(record) unless reflection.nested?`, it does not looks at the result of the `insert_record`. I think this is intended behaviour.

**Comment by dmitry** [comment]
TL;DR Child uniqueness validation should be done only after parent `INSERT`ed.

I have similar issue, and I found that validations with `scope` are done before the parent record is saved, and it checks for `NULL` value, instead of real id value that will be saved only after `INSERT` is completed.

**Comment by dfyx** [comment]
@dmitry not entirely correct as I see it. I have a very similar problem though I create my associations with `build`. For me the order I get in my log file (as far as I can tell from the SQL statements) is the following:

* validate parent
* insert parent
* validate child
* insert child

That looks correct at first glance but there is a problem. Under some circumstances (I can't exactly tell you when but I think it's when I nest my associations even further) the child's validation doesn't see the parent's id even though the parent has already been inserted. If the validation still succeeds the child is inserted with the correct id.

So my guess is that it's just a matter of passing the id to the validator.

**Comment by dmitry** [comment]
@dfyx what version of rails do you have?

**Comment by dfyx** [comment]
Not sure why @dmitry removed his comment but to answer it: I first encountered this on 4.2.1 and it still persisted after I updated the app to 4.2.3.

The thing is, I'm creating a pretty large hierarchy of objects at once (with some of them having associations to existing objects as well) and it only happens under some obscure circumstances that I don't really understand so it's pretty hard to extract a minimal example.

**Comment by dmitry** [comment]
@dfyx I found what rails version you are using through the https://github.com/rails/rails/issues/21058 :)

As I remember, I had an issue with children records, when they are created through the autosave (nested attributes) along with the parent record (rails 4.1.x) I had a problem. For me it was not really important validation, because it should never be invalid, so the database check for my case it just fine. But it would be really nice to support it in the future activerecord versions.

**Comment by dfyx** [comment]
I've found a (really hacky) workaround. I abuse the `if` proc to assign the foreign key.

```
  # Hacky workaround for Rails bug #20676
  validates_uniqueness_of :name, scope: [:project, :parent, :test_plan], if: -> (record) {record.test_plan_id = record.test_plan.id if record.test_plan.present?; true}
 ```

Now everything works as expected.

**Comment by dmitry** [comment]
Looks like an issue somewhere in stale / invert in `Autosave` module.

**Comment by chirag7jain** [comment]
I having this issue on rails 4.2.5..

**Comment by maclover7** [comment]
Can reproduce on master -- the only error I'm seeing is with the last test, where the new `Country` is being created. Would you be willing to do a little bit of some sourcediving and try and figure out where this bug is introduced, and how it can be solved?

**Comment by bluehallu** [comment]
Can reproduce on Rails 5 beta 3

**Comment by lucassch** [comment]
I having this issue on rails 4.2.3

**Comment by dyoganand** [comment]
The same issue persists in rails 4.2.7 as well

**Comment by Jaroost** [comment]
I have the same issue in rails 5.0.0.1. 

There is no other solution than setting a unique index on 2 columns and handle the ActiveRecord::RecordNotUnique exception in the controller?

**Comment by vspar** [comment]
I also have the same issue on rails 5.0.0.1.

Here's my sample code:
`class PeopleGroupItem < ApplicationRecord`
`  belongs_to :people_group`
`  validates :type, presence: true, uniqueness: {scope: :people_group, message: 'combination already used'}`
`  validates :value, presence: true`
`end`

I also tried scope: :group_id but to no avail.

However, the validation of the presence requirement does work. It's just the uniqueness check which is ignored.

**Comment by Jaroost** [comment]
Ok I found a working solution with [this](https://techbrownbags.wordpress.com/2014/02/05/rails-validation-of-cocoon-nested-forms/) link:

If you are using Rails 5+ just add this to ApplicationRecord class
```
 def self.validates_uniqueness(*attr_names)
    # Set the default configuration
    configuration = { attribute_name: :name , message: I18n.t('validations.duplicate') }

    # Update defaults with any supplied configuration values
    configuration.update(attr_names.extract_options!)

    validates_each(attr_names) do |record, record_attr_name, value|
      duplicates = Set.new
      attr_name = configuration[:attribute_name]
      value.map do |obj|
        cur_attr_value = obj.try(attr_name)
        if duplicates.member?(cur_attr_value)
          # mark the record as in error so validation will detect a failure
          record.errors.add(record_attr_name, '')
          obj.errors.add(attr_name, configuration[:message])
        else
          duplicates.add(cur_attr_value)
        end
      end
    end
  end
```
And on the model who hold the accept_nested_attributes do something like that

```
has_many :components, :class_name => 'Machine::Component'
  accepts_nested_attributes_for :components, allow_destroy: true
  validates_uniqueness :components, attribute_name: 'name'
```

**Comment by Jaroost** [comment]
```
 def self.validates_uniqueness(*attr_names)
    # Set the default configuration
    configuration = { attribute_name: :name , message: I18n.t('validations.duplicate') }

    # Update defaults with any supplied configuration values
    configuration.update(attr_names.extract_options!)

    validates_each(attr_names) do |record, record_attr_name, value|
      duplicates = Set.new
      attr_name = configuration[:attribute_name]
      value.map do |obj|
        #be sure to not count the destroyed objects
        unless obj.marked_for_destruction?
          cur_attr_value = obj.try(attr_name)
          if duplicates.member?(cur_attr_value)
            # mark the record as in error so validation will detect a failure
            record.errors.add(record_attr_name, '')
            obj.errors.add(attr_name, configuration[:message])
          else
            duplicates.add(cur_attr_value)
          end
        end
      end
    end
  end
```

Adding the `unless obj.marked_for_destruction?` when you try to destroy and recreate the same unique associated model

**Comment by vspar** [comment]
Very helpful. Thanks.

**Comment by nminhduc** [comment]
What solution for rails 4?

**Comment by Jaroost** [comment]
You can add the method `def self.validates_uniqueness(*attr_names)` to your model you want to check.

You can also create a new class ApplicationRecord in your model that inherit `ActiveRecord::Base` and after add the method `def self.validates_uniqueness(*attr_names)` in that class. Then change the base class of your model to the  ApplicationRecord class

**Comment by goravbhootra** [comment]
Confirming presence of issue in Rails 5.0.1 as well

**Comment by iainbeeston** [comment]
In my case I'm calling `.valid?`, and so the associated models are not saved to the database at all. I still think validation should fail though. I think the only way for this to work would be for the validation to check if the scope refers to an association, and if so validate uniqueness against sibling objects models *in memory*.

For example:

    class Parent < ActiveRecord::Base
      has_many :children
      accepts_nested_attributes_for :children
    end

    class Child < ActiveRecord::Base
      belongs_to :parent
      validates :name, uniqueness: { scope: :parent }
    end

    # I'd expect this to raise a ValidationError
    Parent.new(children_attributes: [{ name: 'John' }, { name: 'John' }]).validate!

In this case I'd imagine that when each Child is validated, the UniquenessValidator would look up `parent.children` and see if any of the other children share the same name.

Looking at the code for[UniquenessValidator](https://github.com/rails/rails/blob/c2b2a8c74fcd2e32996641503ed029c8f7484b00/activerecord/lib/active_record/validations/uniqueness.rb) at the moment it's hard-coded to work using `where()` (see line 95). That's fine when individual models are being created or updated, but won't work when you're creating multiple models at once as we do with accepts nested attributes.

**Comment by arturhaddad** [comment]
I still have this problem in Rails 5.0.5, here is a [post i made in StackOverflow describing with details.](https://stackoverflow.com/questions/46062085/how-to-validate-combined-values-in-rails-5)

**Comment by ZASMan** [comment]
Agreed, there still doesn't seem to be a fix for it. I have a model with a nested form and then I'm also trying to add validations for presence and other data quality in the model thats being nested, but to no avail.

**Comment by ZASMan** [comment]
Should be contributes to rails core ^^

**Comment by phspies** [comment]
@thuhuongly This doesn't seem to work on rails 5.1.5. The "value.map" throws an exception as it's expecting an array but a string is encountered.

**Comment by OKitsMikael** [comment]
@phspies I had the same issue; my problem was that I placed the `validates_uniqueness` method on the child model where you'd normally put the validation. Placing it on the parent, where you declare `accepts_nested_attributes_for`, fixed the issue. Hope that helps!

**Comment by goravbhootra** [comment]
usage example:

There are three models, organization, user and join organization_user. Following needs to be placed in organzation.rb -

```
validates_uniqueness :organization_users, 
                                    attribute_name: 'organization_id',
                                    scope: [:user_id], 
                                    message: 'User is already part of the Organization'
```

**Comment by anka** [comment]
I am also facing this bug within `Rails 5.1.5`. Within [this stackoverflow post](https://stackoverflow.com/questions/46062085/how-to-validate-combined-values-in-rails-5/50106261#50106261) I try to explain my problem and provide a simple solution using a `uniq database index` and a `around_save` callback.

**Comment by muhammad-ahmad-shafiq** [comment]
@armchairdj 
Hi I'm new in ROR.
I am facing a similar problem. I want to ask where to add that custom validator. I am using Rails 4.2.8.

**Comment by armchairdj** [comment]
@muhammad-ahmad-shafiq Validators go in `app/validators`.

https://stackoverflow.com/questions/42253003/where-should-i-put-the-custom-validators-in-rails-5

**Comment by ziaulrehman40** [comment]
This issue is very annoying and still open after years.
@armanddp Can you please confirm if your solution will work in rails 4?

**Comment by Mihoid** [comment]
try use { scope: :model }, not { scope: :model_id }
like this 
`validates :language , uniqueness: { scope: :country }`

it work's for me :)

**Comment by tacchistudios-doug** [comment]
Still happening on Rails 5.2.1; Most workarounds in this topic is pretty hacky. I might as well drop accept `accepts_nested_attributes_for` all together and create the parent and child separately.

**Comment by djezzzl** [comment]
@tacchistudios-doug Hey, did you try [validates_db_uniqueness_of](https://github.com/toptal/database_validations#validates_db_uniqueness_of)?

**Comment by tacchistudios-doug** [comment]
> @tacchistudios-doug Hey, did you try [validates_db_uniqueness_of](https://github.com/toptal/database_validations#validates_db_uniqueness_of)?

No, I did not. I didn't want to add another gem in the project.

**Comment by djezzzl** [comment]
@tacchistudios-doug I understand, but be aware it will fix a problem.

**Comment by matsumura-taichi** [comment]
test

**Comment by cknoxrun** [comment]
Hi @djezzzl it's a great gem, and it works in some cases, but just FYI not for polymorphic belongs_to (obviously since there is no FK). We still need a general solution there.

**Comment by Zyndoras** [comment]
Not that I expected anything else, but the problem continues to exist in Rails 6.0 (beta).

**Comment by mateusdeap** [comment]
Just commenting here to say that I also am facing this issue with rails 5.2.2

**Comment by lethunder** [comment]
hello everyone,
i decided to implement a simple after_validation on the nested model.

Something like in restricted_endpoint model

validates :code, presence: true, uniqueness: { scope: :client }
after_validation :restricted_code_should_be_uniq

private
  def restricted_code_should_be_uniq
    self.errors.add(:code, :invalid) if client.restricted_endpoints.where(code: code)
  end

this is only temporary until the problem is solved

**Comment by dfyx** [comment]
I wouldn't count on your workaround being temporary. This issue has been open for four years and two major releases. I'm not sure if this will ever be fixed.

**Comment by lethunder** [comment]
sure @dfyx. The code is not that uglly. It can also stay like that :)

**Comment by tbsvttr** [comment]
Same issue with Rails 5.2.3 on Ruby 2.6.3.

**Comment by seliverstov-maxim** [comment]
Fix please

**Comment by kbleabres** [comment]
Please fix this

**Comment by anhkhang2304** [comment]
Please fix asap.

**Comment by 1v** [comment]
Not fixed 👍

**Comment by catogonzalez** [comment]
My not-so-hacky solution that might as well be converted to a custom validator if you have to use it in many places:
```
validate :my_attribute_uniqueness

private 

# parent_record: the one declaring the accepts_nested_attributes_for
# nested_collection: the name of the collection as declared in accepts_nested_attributes_for :nested_collection
# my_attribute: the name of the attribute you want to validate with uniqueness: { scope: ... }
# 
def my_attribute_uniqueness
  return if parent_record.nil?
  errors.add(:my_attribute, :taken) if parent_record.nested_collection.any? { |sibling| sibling != self && sibling.my_attribute == self.my_attribute }
end
```

**Comment by kofiasare** [comment]
I am using Rails 6, and updated @armchairdj solution with case sensitivity
```ruby
def self.validate_nested_uniqueness_of(*nested_attrs)
  opts           = nested_attrs.extract_options!
  watch          = opts[:watch]
  case_sensitive = opts[:case_sensitive]
  scope          = opts[:scope    ] || []
  error_key      = opts[:error_key] || :nested_taken
  message        = opts[:message  ] || nil

  raise ArgumentError unless watch.present?
  validates_each(nested_attrs) do |record, nested_attr, nested_values|
    dupes = Set.new
    nested_values.reject(&:marked_for_destruction?).map do |nested_val|
      dupe        = scope.each.inject({}) { |memo, (k)| memo[k] = nested_val.try(k); memo }
      dupe[watch] = nested_val.try(watch)
      dupe[watch] = dupe[watch].try(:downcase)  if case_sensitive == false

      if dupes.member?(dupe)
        record.errors.add(:base, error_key, message: message)
      else
        dupes.add(dupe)
      end
    end
  end
end
```

**Comment by dben89x** [comment]
How is this not fixed yet? Nearly 5 years later...

**Comment by tbsvttr** [comment]
Problem still exists with Rails 6 and Ruby 2.6.5. Please fix this.

**Comment by SylarRuby** [comment]
Bumping in 2020. Any fix?

**Comment by letiesperon** [comment]
Any updates on this? 🙏

**Comment by YassineDM** [comment]
Also interested in an 'official' fix...

**Comment by YassineDM** [comment]
> I am using Rails 6, and updated @armchairdj solution with case sensitivity
> 
> ```ruby
> def self.validate_nested_uniqueness_of(*nested_attrs)
>   opts           = nested_attrs.extract_options!
>   watch          = opts[:watch]
>   case_sensitive = opts[:case_sensitive]
>   scope          = opts[:scope    ] || []
>   error_key      = opts[:error_key] || :nested_taken
>   message        = opts[:message  ] || nil
> 
>   raise ArgumentError unless watch.present?
>   validates_each(nested_attrs) do |record, nested_attr, nested_values|
>     dupes = Set.new
>     nested_values.reject(&:marked_for_destruction?).map do |nested_val|
>       dupe        = scope.each.inject({}) { |memo, (k)| memo[k] = nested_val.try(k); memo }
>       dupe[watch] = nested_val.try(watch)
>       dupe[watch] = dupe[watch].try(:downcase)  if case_sensitive == false
> 
>       if dupes.member?(dupe)
>         record.errors.add(:base, error_key, message: message)
>       else
>         dupes.add(dupe)
>       end
>     end
>   end
> end
> 
> # parent model
> validate_nested_uniqueness_of :fields, watch: :name, scope: [:template_id], case_sensitive: false
> ```

Could you please explain the use case for the `watch` option?

**Comment by EdOverride** [comment]
I think there is a better and easier solution, which is to just add custom validator for new records.

```ruby
class Product < ApplicationRecord
  has_many :variants
  accepts_nested_attributes_for :variants, allow_destroy: true
end
```

```ruby
class Variant < ApplicationRecord
  belongs_to :product

  validates :name,
            presence:   true,
            uniqueness: { case_sensitive: false, scope: :product_id }

  validate :new_nested_name_presence

  def new_nested_name_presence
    return unless self.product.id.nil?

    if self
         .product
         .variants
         .map    { |v| v.name }
         .select { |x| x.downcase == self.name.downcase }
         .length > 1
      self.errors.add :name, I18n.t('errors.messages.taken')
    end
  end

end
```

**Comment by EdOverride** [comment]
I think there is a better and easier solution, which is to just add custom validator for new records.
This way it adds error to the `Variant.name` attribute not just `Product.variants`.

```ruby
class Product < ApplicationRecord
  has_many :variants
  accepts_nested_attributes_for :variants, allow_destroy: true
end
```

```ruby
class Variant < ApplicationRecord
  belongs_to :product

  validates :name,
            presence:   true,
            uniqueness: { case_sensitive: false, scope: :product_id }

  validate :new_nested_name_presence

  def new_nested_name_presence
    return unless self.product.id.nil?

    if self
         .product
         .variants
         .select { |x| x.name.downcase == self.name.downcase }
         .length > 1
      self.errors.add :name, I18n.t('errors.messages.taken')
    end
  end

end
```

**Comment by helenatxu** [comment]
I'm also waiting for a fix to this problem.

**Comment by BSorbus** [comment]
if it is to work universally (also for integer keys) the line:

`dupe[watch] = dupe[watch].try(:downcase)  if case_sensitive == false`

should be changed to 

`dupe[watch] = dupe[watch].try(:downcase)  if case_sensitive == false && dupe[watch].class == 'String'`

**Comment by BSorbus** [comment]
if it is to work universally (also for integer keys) the line:

`dupe[watch] = dupe[watch].try(:downcase)  if case_sensitive == false`

should be changed to 

`dupe[watch] = dupe[watch].try(:downcase)  if case_sensitive == false && dupe[watch].class == 'String'`

**Comment by thomasfedb** [comment]
@rafaelfranca do you have any particular thoughts on the best way to address this? I could contribute a PR given a bit of direction.

**Comment by anubhav8421** [comment]
Verified this on Rails 6.0.3.2. Would be great if this is addressed.

**Comment by intrip** [comment]
[update reproduction script](https://gist.github.com/intrip/25665e7b9e759cb021c951ea4d25f28b)

The key issue is that the validation of the associated record is called before saving the parent record which leads into not finding the uniqueness validation error .
What we could do is to define a set of validations (just uniqueness for now) which should run after the parent record has been saved and if any of such validation exists we should validate also after the save/update and rollback in case of validation error.

I'd like some feedback from the core team about this proposal.

**Comment by sharvy** [comment]
@spurnaye This is how I worked around it:

```
class Country < ActiveRecord::Base
  has_many :cities, dependent: :destroy
  accepts_nested_attributes_for :cities
end

class City < ActiveRecord::Base
  belongs_to :country
  validates :language , uniqueness: { scope: :country_id }

  #
  # This after save callback will make sure that the transaction initiated by accepts_nested_attributes_for aborted
  # if the children are saved in spite of having validation errors.
  #
  after_save :raise_if_saved_with_errors
  def raise_if_saved_with_errors
    raise ActiveRecord::RecordInvalid.new(self) if errors.any?
  end
end
```

**Comment by jish0desarkar** [comment]
> I modify @Jaroost 's solution a little bit to work with **scope**:
> 
> ```
> def self.validates_uniqueness(*attr_names)
>     # Set the default configuration
>     configuration = { attribute_name: :name, scope: [], message: 'has already been taken' }
> 
>     # Update defaults with any supplied configuration values
>     configuration.update(attr_names.extract_options!)
> 
>     validates_each(attr_names) do |record, record_attr_name, value|
>       duplicates = Set.new
>       attr_name = configuration[:attribute_name]
>       scope = configuration[:scope]
>       value.map do |obj|
>         ob = {}
>         ob[attr_name] = obj.try(attr_name)
>         scope.each { |att| ob[att] = obj.try(att) }
>         if duplicates.member?(ob)
>           # mark the record as in error so validation will detect a failure
>           record.errors.add(record_attr_name, '')
>           obj.errors.add(attr_name, configuration[:message])
>         else
>           duplicates.add(ob)
>         end
>       end
>     end
>   end
> ```
> 
> So that you can do something like this on your model:
> 
> `validates_uniqueness :teacher_id, scope: [:semester_id, :class_id]`

I'm getting an undefined method `validates_uniqueness` did you mean `validates_uniqueness_of` error. How to use this method?

**Comment by david-mears-2** [comment]
I've gone with @catogonzalez 's solution, adapted to be reusable across different classes.

```ruby
class NestedUniquenessValidator < ActiveModel::EachValidator
  def validate_each(record, attribute, value)
    if siblings(record).any? { |sibling| sibling.public_send(attribute) == value }
      record.errors.add(attribute, "has already been taken")
    end
  end

  private

  def siblings(record)
    begin
      siblings = record.parent.public_send(sibling_accessor(record))
    rescue NoMethodError
      raise "The record should implement the method `parent`, which returns the associated instance of the class which specifies `accepts_nested_attribute_for [child_class]`"
    end

    siblings - [record] # Using the minus operator since `delete` had undesired side-effects (records were destroyed)
  end

  # Works out the method for accessing the collection of nested objects (children of the parent class) based on the record's class name
  def sibling_accessor(record)
    # Your implementation here (project-dependent)
  end
end
```

I am using this alongside ActiveRecord's validator in the classes in question, since the nested_uniqueness validator doesn't cover cases where records are created *without* using accepts_nested_attributes_for: 

```ruby
class SuperClass < ApplicationRecord
  has_many :sub_classes
  accepts_nested_attributes_for :sub_classes
end

class SubClass < ApplicationRecord
  belongs_to :super_class

  validates :detail, uniqueness: {scope: :super_class}, if: -> { detail.present? && errors[:detail].none? }
  validates :detail, nested_uniqueness: true, if: -> { detail.present? && errors[:detail].none? }

  def parent
    super_class
  end
end
```

**Comment by aka-nez** [comment]
Issue still here in `rails 7.0.4.2`

**Comment by aka-nez** [comment]
I've ended up with this gem https://github.com/toptal/database_validations#rescue-option

```ruby
class Approver < ApplicationRecord
  belongs_to :approvable, polymorphic: true
  belongs_to :user 
  validates_db_uniqueness_of :user_id, scope: %i[approvable_id approvable_type], rescue: :always
end

class Template < ApplicationRecord
  has_many :approvers, -> { order(position: :asc) }, as: :approvable
  accepts_nested_attributes_for :approvers, reject_if: :all_blank, allow_destroy: true
end
```

Works fine with create / update template record

**Comment by KRSMHDK** [comment]
Any updates on this issue?  still figuring this out

**Comment by BSorbus** [comment]
Do you think it would be a good idea to introduce an "allow_blank" parameter?



`  
def validate_nested_uniqueness_of(*nested_attrs)
    opts           = nested_attrs.extract_options!
    uniq_key       = opts[:uniq_key]
    case_sensitive = opts[:case_sensitive]
    scope          = opts[:scope    ] || []
    error_key      = opts[:error_key] || :nested_taken
    message        = opts[:message  ] || nil
    allow_blank    = opts[:allow_blank] || false
    raise ArgumentError unless uniq_key.present?
    validates_each(nested_attrs) do |record, nested_attr, nested_values|
      dupes = Set.new
      nested_values.reject(&:marked_for_destruction?).map do |nested_val|
        dupe = scope.each.inject({}) { |memo, (k)| memo[k] = nested_val.try(k); memo }
        dupe[uniq_key] = nested_val.try(uniq_key)
        dupe[uniq_key] = dupe[uniq_key].try(:downcase) if (case_sensitive == false && dupe[uniq_key].class == 'String')
        
        if dupes.member?(dupe)
          record.errors.add(:base, error_key, message: message) unless allow_blank == false ? false : dupe[uniq_key].blank?
        else
          dupes.add(dupe)
        end

      end 
    end
  end
  `

**Comment by tatcn8** [comment]
I just used @aka-nez solution and it worked fine for me on `7.0.4`

**Comment by JamesAndresCM** [comment]
Hi, this validation apparently solves the problem, but you need to remove `validates :language , uniqueness: { scope: :country_id }`.
I also believe another good idea is to add a unique index in the database and add inverse associations.

````
class City < ApplicationRecord
  belongs_to :country, inverse_of: :cities

  validate :language_must_be_unique_within_country

  def language_must_be_unique_within_country
    return if language.blank? || country.nil?

    siblings = country.cities.reject do |c|
      c == self || c.marked_for_destruction?
    end

    if siblings.any? { |c| c.language == language }
      errors.add(:language, 'has already been taken within this country')
    end
  end
end
````

````
class Country < ApplicationRecord
  has_many :cities, dependent: :destroy, inverse_of: :country
  accepts_nested_attributes_for :cities, allow_destroy: true
end
````
then

````
country = Country.new(description:'abc', language: 'en-US', cities:[City.new(language: 'en-US', text: 'abc'), City.new(language: 'en-US', text: 'abc')]).save!
   (0.2ms)  BEGIN
   (0.2ms)  ROLLBACK
Traceback (most recent call last):
        1: from (irb):3
ActiveRecord::RecordInvalid (Validation failed: Cities language has already been taken within this country)
irb(main):004:0> country = Country.new(language: "es").save
   (0.2ms)  BEGIN
  Country Create (0.9ms)  INSERT INTO "countries" ("language", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["language", "es"], ["created_at", "2025-07-19 23:20:36.099672"], ["updated_at", "2025-07-19 23:20:36.099672"]]
   (0.9ms)  COMMIT
=> true
irb(main):005:0> city1 = Country.first.cities.create(language: "es")
  Country Load (0.6ms)  SELECT  "countries".* FROM "countries" ORDER BY "countries"."id" ASC LIMIT $1  [["LIMIT", 1]]
   (0.4ms)  BEGIN
  City Load (0.5ms)  SELECT "cities".* FROM "cities" WHERE "cities"."country_id" = $1  [["country_id", 11]]
  City Create (0.4ms)  INSERT INTO "cities" ("language", "country_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["language", "es"], ["country_id", 11], ["created_at", "2025-07-19 23:20:51.987812"], ["updated_at", "2025-07-19 23:20:51.987812"]]
   (0.5ms)  COMMIT
=> #<City id: 17, language: "es", text: nil, country_id: 11, created_at: "2025-07-19 23:20:51", updated_at: "2025-07-19 23:20:51">

irb(main):007:0> country = Country.first
  Country Load (1.8ms)  SELECT  "countries".* FROM "countries" ORDER BY "countries"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> #<Country id: 11, description: nil, language: "es", created_at: "2025-07-19 23:20:36", updated_at: "2025-07-19 23:20:36">
irb(main):008:0> country.update!(cities_attributes: [{id: 17, _destroy: true}, {language: "es"}])
   (0.2ms)  BEGIN
  City Load (0.4ms)  SELECT "cities".* FROM "cities" WHERE "cities"."country_id" = $1 AND "cities"."id" = $2  [["country_id", 11], ["id", 17]]
  City Load (0.2ms)  SELECT "cities".* FROM "cities" WHERE "cities"."country_id" = $1  [["country_id", 11]]
  City Destroy (0.2ms)  DELETE FROM "cities" WHERE "cities"."id" = $1  [["id", 17]]
  City Create (0.2ms)  INSERT INTO "cities" ("language", "country_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"  [["language", "es"], ["country_id", 11], ["created_at", "2025-07-19 23:21:19.906962"], ["updated_at", "2025-07-19 23:21:19.906962"]]
   (0.5ms)  COMMIT
=> true
````

**Comment by mamantoha** [comment]
By the way, if anyone’s interested. I maintain a repository [validates_nested_uniqueness](https://github.com/mamantoha/validates_nested_uniqueness)￼. It validates the uniqueness of associated records when using `accepts_nested_attributes_for`.

We’ve been using it in production for years, and it’s proven to be stable and reliable.
